### PR TITLE
Respect "grpc-status-details-bin" trailer on error response

### LINKED ISSUE #49196: [gRPC] How to correctly transmit error details to client with gRPC-JSON transcoding?
### Discussed in https://github.com/dotnet/aspnetcore/discussions/49179

<div type='discussions-op-text'>

<sup>Originally posted by **Sayan751** July  3, 2023</sup>
I have a gRPC service with JSON transcoding enabled. An endpoint of the service throws `RpcException` with same status code but different metadata. Examples:

```csharp
throw new RpcException(new Status(StatusCode.InvalidArgument, "Whatever1"), new Metadata
{
    { "Code", Errors.Error1.ToString() }
});

throw new RpcException(new Status(StatusCode.InvalidArgument, "Whatever2"), new Metadata
{
    { "Code", Errors.Error2.ToString() }
});
```

Whoever, when I am making an HTTP call to the service, then I am getting the following:

```json
{
  "code": 3,
  "message": "Whatever1",
  "details": []
}
```

I expected the trailers to appear in the `details` array. Am I missing something here? At the end, I need a way to differentiate between the two errors with the same status code. Is that possible? </div>

### PR DESCRIPTION
**Pull request by alrz**
Fixes https://github.com/dotnet/aspnetcore/issues/49196

### CODE REVIEW

**Comment by adityamandaleeka** [comment]
@JamesNK PTAL

**Comment by alrz** [comment]
@JamesNK for review. Thanks.

File: src/Grpc/JsonTranscoding/src/Microsoft.AspNetCore.Grpc.JsonTranscoding/Internal/JsonTranscodingServerCallContext.cs [hunk 1]
```diff
@@ -109,6 +110,7 @@ internal async Task ProcessHandlerErrorAsync(Exception ex, string method, bool i
             GrpcServerLog.RpcConnectionError(Logger, rpcException.StatusCode, rpcException.Status.Detail, rpcException.Status.DebugException);
 
             status = rpcException.Status;
+            trailers = rpcException.Trailers;
```
Comment (JamesNK) [inline src/Grpc/JsonTranscoding/src/Microsoft.AspNetCore.Grpc.JsonTranscoding/Internal/JsonTranscodingServerCallContext.cs:113]: Add the trailers from the exception to the `ResponseTrailers` collection (a property on this type) and then use it.

e.g. https://github.com/grpc/grpc-dotnet/blob/1625f8942791c82d700802fc7278c543025f0fd3/src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs#L187-L190

File: src/Grpc/JsonTranscoding/src/Microsoft.AspNetCore.Grpc.JsonTranscoding/Internal/JsonTranscodingServerCallContext.cs [hunk 2]
```diff
@@ -119,9 +121,10 @@ internal async Task ProcessHandlerErrorAsync(Exception ex, string method, bool i
             // Note that the exception given to status won't be returned to the client.
             // It is still useful to set in case an interceptor accesses the status on the server.
             status = new Status(StatusCode.Unknown, message, ex);
+            trailers = null;
```
Comment (JamesNK) [inline src/Grpc/JsonTranscoding/src/Microsoft.AspNetCore.Grpc.JsonTranscoding/Internal/JsonTranscodingServerCallContext.cs:124]: Use `ResponseTrailers`.

File: src/Grpc/JsonTranscoding/test/Microsoft.AspNetCore.Grpc.JsonTranscoding.Tests/UnaryServerCallHandlerTests.cs [hunk 3]
```diff
@@ -871,6 +871,81 @@ public async Task HandleCallAsync_RpcExceptionThrown_StatusReturned()
         Assert.Equal(debugException, exceptionWrite.Exception);
     }
 
+    [Fact]
+    public async Task HandleCallAsync_RpcExceptionThrown_StatusDetailsReturned()
```
Comment (JamesNK) [inline src/Grpc/JsonTranscoding/test/Microsoft.AspNetCore.Grpc.JsonTranscoding.Tests/UnaryServerCallHandlerTests.cs:875]: Duplicate this test but set `grpc-status-details-bin` on the context's `ResponseTrailers` collection. Throw a non-gRPC exception like `InvalidOperationException`.

File: src/Grpc/JsonTranscoding/src/Microsoft.AspNetCore.Grpc.JsonTranscoding/Internal/JsonRequestHelpers.cs [hunk 4]
```diff
@@ -82,21 +82,33 @@ public static (Stream stream, bool usesTranscodingStream) GetStream(Stream inner
...
 
         await WriteResponseMessage(response, encoding, e, options, CancellationToken.None);
+
+        static Google.Rpc.Status? GetStatusDetails(Metadata? trailers)
+        {
+            const string GrpcStatusDetailsHeader = "grpc-status-details-bin";
+            var statusDetails = trailers?.Get(GrpcStatusDetailsHeader);
+            if (statusDetails != null && statusDetails.IsBinary)
+            {
+                return Google.Rpc.Status.Parser.ParseFrom(statusDetails.ValueBytes);
```
Comment (JamesNK) [inline src/Grpc/JsonTranscoding/src/Microsoft.AspNetCore.Grpc.JsonTranscoding/Internal/JsonRequestHelpers.cs:107]: Make it more obvious what has gone wrong if there is invalid content in `grpc-status-details-bin`.
```suggestion
                try
                {
                    return Google.Rpc.Status.Parser.ParseFrom(statusDetails.ValueBytes);
                }
                catch (Exception ex)
                {
                    throw new InvalidOperationException(@"Error when parsing the ""grpc-status-details-bin"" trailer.", ex);
                }
```

**Review by JamesNK — COMMENTED** [review]
Looks good. There are some changes required before merging. The main one is support for `ServerCallContext.ResponseTrailers`.

File: src/Grpc/JsonTranscoding/src/Microsoft.AspNetCore.Grpc.JsonTranscoding/Internal/JsonTranscodingServerCallContext.cs (hunk 1 above)
Comment (alrz) [reply src/Grpc/JsonTranscoding/src/Microsoft.AspNetCore.Grpc.JsonTranscoding/Internal/JsonTranscodingServerCallContext.cs:113]: @JamesNK ResponseTrailersCore currently throws NIE; did you mean to hold this off until it's implemented?

File: src/Grpc/JsonTranscoding/src/Microsoft.AspNetCore.Grpc.JsonTranscoding/Internal/JsonTranscodingServerCallContext.cs [hunk 5]
```diff
@@ -164,7 +168,7 @@ protected override Metadata RequestHeadersCore
 
     protected override CancellationToken CancellationTokenCore => HttpContext.RequestAborted;
 
-    protected override Metadata ResponseTrailersCore => throw new NotImplementedException();
+    protected override Metadata ResponseTrailersCore { get; } = new();
```
Comment (alrz) [inline src/Grpc/JsonTranscoding/src/Microsoft.AspNetCore.Grpc.JsonTranscoding/Internal/JsonTranscodingServerCallContext.cs:171]: Not sure if this is supposed to end up in response headers.

**Review by JamesNK — APPROVED** [review]

**Comment by JamesNK** [comment]
Thanks!

**Comment by alrz** [comment]
@JamesNK I'm not familar with how PRs are triaged here, could this kind of fix possibly make it to a point release?

**Comment by JamesNK** [comment]
No, this is a new feature. It will come in .NET 9

**Pull request merged**